website/docs/script/type-id.mdx

```diff
@@ -0,0 +1,237 @@
+
+Note that this question might seem irrelevant to Script upgradability right now, but please bear with me, we will see how it will contribute to the final solution in CKB.
+
+The answer is quite simple, here's the basic workflow of the Script:
+
+1. Count how many output cells use current type Script, if there's more than one cell with current type Script, returns a failure return code;
+2. Count how many input cells use current type Script, if there's one input cell with current type Script, returns a success return code;
+3. Use CKB syscall to read the first OutPoint in current transaction;
+4. If the OutPoint data read match the `args` part of current type Script, returns a success return code;
+5. Returns a failure return code otherwise.
+
+Putting in simple C code, the Script would look like following:
+
+```c
+#include "blockchain.h"
+#include "ckb_syscalls.h"
+
+#define INPUT_SIZE 128
+#define SCRIPT_SIZE 32768
+
+int main() {
+  uint64_t len = 0;
+  int ret = ckb_load_cell(NULL, &len, 0, 1, CKB_SOURCE_GROUP_OUTPUT);
+  if (ret != CKB_INDEX_OUT_OF_BOUND) {
+    return -1;  /* 1 */
+  }
+
+  len = 0;
+  ret = ckb_load_cell(NULL, &len, 0, 0, CKB_SOURCE_GROUP_INPUT);
+  if (ret != CKB_INDEX_OUT_OF_BOUND) {
+    return 0;  /* 2 */
+  }
+
+  /* 3 */
+  unsigned char input[INPUT_SIZE];
+  uint64_t input_len = INPUT_SIZE;
+  ret = ckb_load_input(input, &input_len, 0, 0, CKB_SOURCE_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (input_len > INPUT_SIZE) {
+    return -100;
+  }
+
+  unsigned char Script[SCRIPT_SIZE];
+  len = SCRIPT_SIZE;
+  ret = ckb_load_script(Script, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (len > SCRIPT_SIZE) {
+    return -101;
+  }
+  mol_seg_t script_seg;
+  script_seg.ptr = (uint8_t *)Script;
+  script_seg.size = len;
+
+  if (MolReader_Script_verify(&script_seg, false) != MOL_OK) {
+    return -102;
+  }
+
+  mol_seg_t args_seg = MolReader_Script_get_args(&script_seg);
+  mol_seg_t args_bytes_seg = MolReader_Bytes_raw_bytes(&args_seg);
+
+  if ((input_len == args_bytes_seg.size) &&
+      (memcmp(args_bytes_seg.ptr, input, input_len) == 0)) {
+    /* 4 */
+    return 0;
+  }
+  return -103;
+}
+```
+
+Attackers will be prevented in several different directions:
+
+1. If an attacker tries to create a cell with the exactly same type Script, there will be 2 cases:
+   a. A valid transaction will have different OutPoint data in the first input from the type Script args;
+   b. If the user tries to duplicate type Script args as the first transaction input, CKB will signal a double-spent error;
+2. If the attacker tries to use a different type Script args, it will be a different type Script by definition.
+
+This way, we can ensure a cell will have unique type Script across all live cells in CKB. Considering each Script has an associated hash, we will have a cell in CKB with its unique hash, or unique ID.
```

Comment:
```suggestion
Attackers will be prevented in several different ways:

1. If an attacker tries to create a Cell with the exact same Type Script, there will be 2 cases:
   a. A valid transaction will have different OutPoint data in the first input from the Type Script args;
   b. If the user tries to duplicate the Type Script args as the first transaction input, CKB will signal a double-spent error;
2. If the attacker tries to use a different Type Script args, it will be a different Type Script by definition.

This way, we can ensure a Cell will have a unique Type Script across all Live Cells in CKB. Considering each Script has an associated hash, we will have a Cell in CKB with its unique hash, or unique ID.
```